stdlib/LinearAlgebra/src/symmetric.jl

```diff
@@ -856,3 +875,49 @@ for func in (:log, :sqrt)
         end
     end
 end
+
+"""
+    hermitianpart(A, uplo=:U) -> Hermitian
+
+Return the Hermitian part of the square matrix `A`, defined as `(A + A') / 2`, as a
+[`Hermitian`](@ref) matrix. For real matrices `A`, this is also known as the symmetric part
+of `A`. The optional argument `uplo` controls the corresponding argument of the
+[`Hermitian`](@ref) view. For real matrices, the latter is equivalent to a
+[`Symmetric`](@ref) view.
+
+See also [`hermitianpart!`](@ref) for the corresponding in-place operation.
+
+!!! compat "Julia 1.10"
+    This function requires Julia 1.10 or later.
+"""
+hermitianpart(A::AbstractMatrix{T}, uplo::Symbol=:U) where {T} =
+    hermitianpart!(copy_similar(A, Base.promote_op(/, T, Int)), uplo)
+hermitianpart(x::Number, uplo::Symbol=:U) = real(x)
+
+"""
+    hermitianpart!(A, uplo=:U) -> Hermitian
+
+Overwrite the square matrix `A` with its Hermitian part `(A + A') / 2`, and return
+[`Hermitian(A, uplo)`](@ref). For real matrices `A`, this is also known as the symmetric
+part of `A`.
+
+See also [`hermitianpart`](@ref).
+
+!!! compat "Julia 1.10"
+    This function requires Julia 1.10 or later.
+"""
+hermitianpart!(A::AbstractMatrix, uplo::Symbol=:U) =
+    Hermitian(_hermorsympart!(a -> hermitianpart(a, uplo), adjoint, A), uplo)
```

Comment:
since `hermitianpart` returns `Hermitian`, this doesn't seem quite right for the `real` function — if `A` is an array of matrices, it will make the resulting array heterogeneous.

Reply:
Not sure, but here is the output of the corresponding test:

```julia
julia> using LinearAlgebra

julia> A = [randn(ComplexF64, 2, 2) for _ in 1:2, _ in 1:2]
2×2 Matrix{Matrix{ComplexF64}}:
 [0.715485+0.129067im 1.15623-0.781759im; -0.669501-0.666363im 1.37915-0.464405im]  [-1.61413-0.147513im 0.465935-0.0184965im; -0.599647-0.0421609im -0.133311-0.705001im]
 [-0.813572-1.15968im 0.728821+0.241281im; -0.41575-0.572121im 0.56458+0.06045im]   [0.024946+0.998568im 0.76679+0.153056im; -1.1199-0.475701im -0.123403-0.526965im]

julia> hermitianpart(A)
2×2 Hermitian{AbstractMatrix, Matrix{Matrix{ComplexF64}}}:
 ComplexF64[0.715485+0.0im 0.243366-0.0576977im; 0.243366+0.0576977im 1.37915+0.0im]             ComplexF64[-1.21385+0.506083im 0.0250925+0.276812im; 0.0645868-0.141721im 0.215634-0.382725im]
 ComplexF64[-1.21385-0.506083im 0.0645868+0.141721im; 0.0250925-0.276812im 0.215634+0.382725im]  ComplexF64[0.024946+0.0im -0.176555+0.314379im; -0.176555-0.314379im -0.123403+0.0im]

julia> Hermitian((A + A')/2)
2×2 Hermitian{AbstractMatrix, Matrix{Matrix{ComplexF64}}}:
 ComplexF64[0.715485+0.0im 0.243366-0.0576977im; 0.243366+0.0576977im 1.37915+0.0im]             ComplexF64[-1.21385+0.506083im 0.0250925+0.276812im; 0.0645868-0.141721im 0.215634-0.382725im]
 ComplexF64[-1.21385-0.506083im 0.0645868+0.141721im; 0.0250925-0.276812im 0.215634+0.382725im]  ComplexF64[0.024946+0.0im -0.176555+0.314379im; -0.176555-0.314379im -0.123403+0.0im]

julia> Hermitian((A + A')/2) == hermitianpart(A)
true

julia> hermitianpart(A).data == (A + A') / 2
true

julia> typeof(hermitianpart(A).data)
Matrix{Matrix{ComplexF64}} (alias for Array{Array{Complex{Float64}, 2}, 2})
```

I think it doesn't get inhomogeneous because it fills a pre-allocated matrix, and converts the `Hermitian` elements to `Matrix` upon `setindex!`.

Reply:
That seems suboptimal.  Seems clearer (and faster) to me to have something like:
```jl
hermitianpart!(A::AbstractMatrix, uplo::Symbol=:U) = Hermitian(_hermitianpart!(A), uplo)

_hermitianpart(A) = _hermitianpart!(copy(A))
_hermitianpart(a::Number) = real(A)

function _hermitianpart!(A)
    require_one_based_indexing(A)
    n = checksquare(A)
    @inbounds for j in 1:n
        A[j, j] = _hermitianpart(A[j, j])
        for i in 1:j-1
            A[i, j] = val = (A[i, j] + adjoint(A[j, i])) / 2
            A[j, i] = adjoint(val)
        end
    end
    return A
end
```
and then ditch the old `_hermorsympart!` since we no longer have `symmetricpart!`.

Reply:
Included.